src/generics.rs

```diff
@@ -77,6 +77,61 @@ ast_enum_of_structs! {
     }
 }
 
+impl Generics {
+    pub fn type_params(&self) -> TypeParams {
+        TypeParams(self.params.iter())
+    }
+
+    pub fn type_params_mut(&mut self) -> TypeParamsMut {
+        TypeParamsMut(self.params.iter_mut())
+    }
+}
+
+/// An immutable iterator over the type parameters of `Generics`.
+///
+/// Returned by `Generics::type_params`
+pub struct TypeParams<'a>(Iter<'a, GenericParam, Token![,]>);
+
+impl<'a> Iterator for TypeParams<'a> {
+    type Item = &'a TypeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // FIXME: Remove this when ? on Option is stable
+        let next = match self.0.next() {
+            Some(item) => item,
+            None => return None,
+        };
+        if let GenericParam::Type(ref type_param) = *next {
+            Some(type_param)
+        } else {
+            self.next()
+        }
+    }
+}
+
+/// A mutable iterator over the type parameters of `Generics`.
+///
+/// Returned by `Generics::type_params_mut`
+pub struct TypeParamsMut<'a>(IterMut<'a, GenericParam, Token![,]>);
+
+impl<'a> Iterator for TypeParamsMut<'a> {
+    type Item = &'a mut TypeParam;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // FIXME: Remove this when ? on Option is stable
```

Comment:
Syn supports rust versions back to 0.15, so unless we increase our minimum rust version, we won't be able to switch to using ? on Option when it becomes stable.

Reply:
In that case #324 needs to be reverted, since it doesn't compile on 1.15. It's probably worth adding 1.15 to CI if the intent is to continue to support it in all future versions. (I personally don't think it's worth going much farther than the rust-lang-nursery policy)

Reply:
It is being tested in CI (https://github.com/dtolnay/syn/blob/03a092060471ed8bddf492c537d980a57b8f511d/.travis.yml#L18-L19), but travis was choking & I foolishly merged without letting the CI finish.

@dtolnay is the person who has picked our minimum version. I think that our version minimum is 0.15 to match serde's version minimum? Can't remember exactly.

I'll fix up the code to build locally.

Reply:
You mean 1.15 not 0.15, right?

Reply:
yes, 1.15 not 0.15... oops.

Reply:
Fixed in 5680dcc26cbfce6b8590e898df884f47d03c8bed